src/twisted/trial/_asynctest.py

```diff
@@ -190,22 +190,24 @@ def _ebDeferTearDown(self, failure, result):
             result.stop()
         self._passed = False
 
+    @defer.inlineCallbacks
     def deferRunCleanups(self, ignored, result):
         """
         Run any scheduled cleanups and report errors (if any to the result
         object.
         """
-        d = self._runCleanups()
-        d.addCallback(self._cbDeferRunCleanups, result)
-        return d
+        failures = []
+        for func, args, kwargs in self._cleanups[::-1]:
```

Comment:
```suggestion
        for func, args, kwargs in reversed(self._cleanups):
```

Reply:
Note that `reversed()` does not make a copy, but since there is no `removeCleanup()` method, the `_cleanups` list shouldn't be modified during iteration.

Reply:
Does it make sense to update `_runCleanups()` in `SynchronousTestCase` (superclass to the async `TestCase`) to use the same iteration style? Currently it still uses `while` + `pop()`.

Reply:
that allows re-entrant cleanups

Reply:
would it be better to allow re-entrant cleanup in `_asynctest.py`?

Reply:
I'm not sure if re-entrant cleanups are a feature worth having, but if it's a feature that Trial has right now (either by design or by accident), the safest thing is to keep it.

Reply:
what happens if you addCleanup while iterating over reversed(list(...))?

Reply:
With just `reversed(self._cleanups)` it's undefined, with `reversed(list(self._cleanups))` the added cleanup would be ignored.

Reply:
I think sticking with the `self._cleanups[::-1]:` is best as that's idiomatic for reversing and copying a list

Reply:
OK.